src/util/strutil.c

```diff
@@ -8,31 +8,57 @@
 #include <ctype.h>
 #include "RG.h"
 #include "rmalloc.h"
+#include "../../build/icu/build/include/unicode/ustring.h"
 
-void str_tolower(const char *str, char *lower, size_t *lower_len) {
+static void str_changecase(const char *str, char *result, size_t *result_len, const char *locale, bool toUpper) {
 	size_t str_len = strlen(str);
 	//Avoid overflow
-	ASSERT(*lower_len >= str_len);
+	assert(*result_len >= str_len);
+	//Update result len
+	*result_len = str_len;
 
-	//Update lower len
-	*lower_len = str_len;
+	size_t u_og_len = str_len;
+	size_t u_result_len = str_len;
 
-	size_t i = 0;
-	for(; i < str_len; i++) lower[i] = tolower(str[i]);
-	lower[i] = 0;
-}
+	UChar* original = (UChar *) rm_malloc((u_og_len + 1) * sizeof(UChar));
+    UChar* converted = (UChar *) rm_malloc((u_result_len + 1) * sizeof(UChar));
+    UErrorCode errorCode = U_ZERO_ERROR;
 
-void str_toupper(const char *str, char *upper, size_t *upper_len) {
-	size_t str_len = strlen(str);
-	//Avoid overflow
-	ASSERT(*upper_len >= str_len);
+    u_uastrcpy(original, str);
+
+    size_t size;
+	if (toUpper) {
+		size = u_strToUpper(converted, u_result_len, original, u_og_len, locale, &errorCode);
+	} else {
+		size = u_strToLower(converted, u_result_len, original, u_og_len, locale, &errorCode);
+	}
 
-	//Update lower len
-	*upper_len = str_len;
+    // if the result is larger than allocated length
+    if (size > u_result_len) {
```

Comment:
Can you give an example for when `size > u_result_len` is true?

Reply:
The `u_result_len` is equal to the original string length. German `ß` to upper case is `SS` and it is larger.

Reply:
Some characters are encoded differently when they change case. German `ß` to upper case becomes `SS` and it is larger.

I tested it with this code 
```c
UErrorCode errorCode = U_ZERO_ERROR;
UChar o[100];
UChar new[100];

char ochar[100];
char newchar[100];
u_uastrcpy(o, "ß");
u_austrcpy(ochar, o);
u_strToUpper(new, 100, o, 1, "", &errorCode);
u_austrcpy(newchar, new);
printf("uchar length: %s, %d \n", ochar, u_strlen(o));
printf("uchar length: %s, %d \n", newchar, u_strlen(new));
```
and the result is 
```
uchar length: ß, 1 
uchar length: SS, 2
```